File: ai-standard/reference/python/bippsi_unified.py
#!/usr/bin/env python3
"""
Bippsi AI Standard — reference generator (Python)

Produces a conforming /bippsi-unified.md markdown document from a configuration dict.
Dependency-free: requires only Python 3.10+ standard library. No pip install.

Two ways to use:

  1. IMPORT + CALL (dynamic, use in Flask/FastAPI/Django handler):

       from bippsi_unified import render_bippsi_unified

       @app.get("/bippsi-unified.md")
       def unified():
           return Response(render_bippsi_unified(config), media_type="text/markdown")

  2. CLI GENERATION (static):

       python bippsi_unified.py config.json > public/bippsi-unified.md

See SPEC at ../../SPEC.md or
https://github.com/bigappstudiollc/bippsi/blob/main/ai-standard/SPEC.md

License: MIT (see ../../../LICENSE-CODE)
"""

from __future__ import annotations

import json
import sys
from datetime import datetime, timezone
from pathlib import Path
from typing import Any


def render_bippsi_unified(config: dict[str, Any]) -> str:
    """Render a conforming /bippsi-unified.md from a configuration dict.

    Required top-level keys:
      site_url        : 'https://example.com'
      identity        : dict with name, tagline, description, contact, same_as
      agent_policy    : dict with allowed, denied, auth_methods, rate_limits
      mcp             : None | dict with endpoint, protocol_version, transport, tools
      api             : None | str path to openapi.json | list of endpoint dicts
      certification   : None | dict
      licensing       : list of bullet strings
      pages           : list of dicts with url + title
      fallback_files  : list of dicts with path + purpose (optional; defaults to spec v1.0 list)

    See examples/full.md in this repo for a complete config example.
    """
    site_url = (config.get("site_url") or "https://example.com").rstrip("/")
    now = datetime.now(timezone.utc).isoformat(timespec="seconds")
    parts: list[str] = []

    # Header
    parts.append("# Bippsi Unified Agent Manifest\n")
    parts.append("> **Spec:** Bippsi AI Standard v1.0 · <https://github.com/bigappstudiollc/bippsi/blob/main/ai-standard/SPEC.md>")
    parts.append(f"> **Generated:** {now}\n")
    parts.append(
        "> **For AI agents:** this single file replaces the need to fetch `llms.txt`, `AGENTS.md`, "
        "`agents.json`, `openapi.json`, and `/.well-known/mcp.json` separately. "
        "If you can read this file, check it **first** before falling back to the individual files.\n"
    )
    parts.append("---\n")

    # 1. Identity
    parts.append("## 1. Identity\n")
    parts.append("```json")
    parts.append(_json(config.get("identity", {})))
    parts.append("```\n")

    # 2. Agent access policy
    policy = config.get("agent_policy") or {}
    parts.append("## 2. Agent access policy\n")
    if policy.get("allowed"):
        parts.append("**Without authentication, agents MAY:**")
        for item in policy["allowed"]:
            parts.append(f"- {item}")
        parts.append("")
    if policy.get("denied"):
        parts.append("**Without authentication, agents MAY NOT:**")
        for item in policy["denied"]:
            parts.append(f"- {item}")
        parts.append("")
    if policy.get("auth_methods"):
        parts.append("**Authentication methods:**")
        for item in policy["auth_methods"]:
            parts.append(f"- {item}")
        parts.append("")
    if policy.get("rate_limits"):
        parts.append("**Rate limits:**")
        parts.append("```json")
        parts.append(_json(policy["rate_limits"]))
        parts.append("```\n")

    # 3. MCP server
    parts.append("## 3. MCP (Model Context Protocol) server\n")
    if not config.get("mcp"):
        parts.append("This site does not currently run an MCP server.")
        parts.append("Agents should fall through to the REST API (section 4) or legacy discovery (section 8).\n")
    else:
        parts.append("```json")
        parts.append(_json(config["mcp"]))
        parts.append("```\n")
        tools = config["mcp"].get("tools") or []
        if tools:
            parts.append("### Available tools\n")
            for tool in tools:
                parts.append(f"#### `{tool.get('name', '?')}`\n")
                parts.append(f"{(tool.get('description') or '').strip()}\n")

    # 4. REST API
    parts.append("## 4. REST API\n")
    api = config.get("api")
    if not api:
        parts.append("This site does not publish a machine-readable REST API.\n")
    else:
        endpoints = _load_openapi(api) if isinstance(api, str) else api
        parts.append(f"OpenAPI spec: {site_url}/openapi.json\n")
        parts.append("```json")
        parts.append(_json(endpoints))
        parts.append("```\n")

    # 5. Certification
    parts.append("## 5. Certification status (this site)\n")
    if not config.get("certification"):
        parts.append("This site is not currently in any public certification directory.\n")
    else:
        parts.append("```json")
        parts.append(_json(config["certification"]))
        parts.append("```\n")

    # 6. Licensing
    parts.append("## 6. Content licensing for AI training\n")
    licensing = config.get("licensing") or []
    if licensing:
        for line in licensing:
            parts.append(f"- {line}")
        parts.append("")
    else:
        parts.append("No explicit licensing declared. Agents SHOULD NOT assume permission to use content for training or redistribution.\n")

    # 7. Pages
    parts.append("## 7. Page index\n")
    for page in config.get("pages") or []:
        url = page.get("url", "/")
        full = url if url.startswith("http") else f"{site_url}{url}"
        parts.append(f"- [{page.get('title', url)}]({full})")
    parts.append("")

    # 8. Fallback files
    parts.append("## 8. Fallback — individual discovery files\n")
    parts.append("If an agent cannot parse this unified manifest, fall back to the following files at this site's root.")
    parts.append("**If this manifest disagrees with any of them, this manifest is authoritative.**\n")
    parts.append("| File | Purpose |")
    parts.append("|---|---|")
    for f in (config.get("fallback_files") or _default_fallback()):
        parts.append(f"| `{f.get('path', '?')}` | {f.get('purpose', '')} |")
    parts.append("")

    # 9. Spec metadata
    parts.append("## 9. Spec metadata\n")
    parts.append("```json")
    parts.append(_json({
        "spec": "Bippsi AI Standard",
        "spec_version": "1.0",
        "spec_url": "https://github.com/bigappstudiollc/bippsi/blob/main/ai-standard/SPEC.md",
        "spec_source": "https://bippsi.com/bippsi-standard",
        "manifest_generated_at": now,
        "manifest_version": "1.0",
        "publisher": (config.get("identity") or {}).get("name", ""),
    }))
    parts.append("```\n")

    parts.append("---\n")
    parts.append("*This manifest follows the Bippsi AI Standard v1.0. Spec: <https://github.com/bigappstudiollc/bippsi>.*\n")

    return "\n".join(parts)


def _json(data: Any) -> str:
    return json.dumps(data, indent=2, ensure_ascii=False)


def _load_openapi(path: str) -> list[dict[str, str]]:
    """Parse an OpenAPI 3.x spec file and return a compact endpoint list."""
    try:
        spec = json.loads(Path(path).read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return []
    base = (spec.get("servers") or [{}])[0].get("url", "").rstrip("/")
    out = []
    for p, ops in (spec.get("paths") or {}).items():
        for method, op in ops.items():
            if method.lower() not in {"get", "post", "put", "patch", "delete"}:
                continue
            out.append({
                "method":  method.upper(),
                "url":     base + p,
                "summary": op.get("summary", ""),
                "auth":    "Bearer required" if op.get("security") else "none",
            })
    return out


def _default_fallback() -> list[dict[str, str]]:
    """Default fallback-files table per spec v1.0 section 8."""
    return [
        {"path": "/AGENTS.md",                  "purpose": "Narrative agent policy"},
        {"path": "/llms.txt",                   "purpose": "LLM-friendly content index"},
        {"path": "/agents.json",                "purpose": "Structured agent policy"},
        {"path": "/openapi.json",               "purpose": "REST API spec (OpenAPI 3.1)"},
        {"path": "/.well-known/mcp.json",       "purpose": "MCP server manifest"},
        {"path": "/.well-known/ai-plugin.json", "purpose": "OpenAI plugin manifest"},
        {"path": "/.well-known/security.txt",   "purpose": "Security disclosure (RFC 9116)"},
        {"path": "/manifest.json",              "purpose": "PWA manifest"},
        {"path": "/robots.txt",                 "purpose": "Crawler policy"},
        {"path": "/sitemap.xml",                "purpose": "URL index"},
    ]


# ── CLI mode ──────────────────────────────────────────────────────────────────
if __name__ == "__main__":
    if len(sys.argv) < 2:
        sys.stderr.write("Usage: python bippsi_unified.py <config.json> > output.md\n")
        sys.exit(1)
    try:
        cfg = json.loads(Path(sys.argv[1]).read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as e:
        sys.stderr.write(f"Error reading config: {e}\n")
        sys.exit(1)
    sys.stdout.write(render_bippsi_unified(cfg))

File: ai-standard/reference/node/bippsi-unified.js
#!/usr/bin/env node
/**
 * Bippsi AI Standard — reference generator (Node.js)
 *
 * Produces a conforming /bippsi-unified.md markdown document from a configuration object.
 * Dependency-free: requires only Node.js 20+ standard library. No npm install.
 *
 * Two ways to use:
 *
 *   1. IMPORT + CALL (dynamic, use in an Express/Fastify/Hono handler):
 *
 *        import { renderBippsiUnified } from './bippsi-unified.js';
 *        app.get('/bippsi-unified.md', (req, res) => {
 *          res.type('text/markdown').send(renderBippsiUnified(config));
 *        });
 *
 *   2. CLI GENERATION (static):
 *
 *        node bippsi-unified.js config.json > public/bippsi-unified.md
 *
 * See SPEC at ../../SPEC.md or https://github.com/bigappstudiollc/bippsi/blob/main/ai-standard/SPEC.md
 *
 * License: MIT (see ../../../LICENSE-CODE)
 */

import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { argv } from 'node:process';

/**
 * Render a conforming /bippsi-unified.md from a configuration object.
 *
 * Required top-level keys:
 *   site_url         : 'https://example.com'
 *   identity         : { name, tagline, description, contact: { email, url }, same_as: [] }
 *   agent_policy     : { allowed: [], denied: [], auth_methods: [], rate_limits: {} }
 *   mcp              : null | { endpoint, protocol_version, transport, tools: [{name, description}] }
 *   api              : null | path/to/openapi.json | array of endpoint objects
 *   certification    : null | { certified, overall_score, status, verify_url, ... }
 *   licensing        : [line, line, ...]  (bullet strings)
 *   pages            : [{ url, title }]
 *   fallback_files   : [{ path, purpose }]
 *
 * See examples/full.md in this repo for a complete config.
 */
export function renderBippsiUnified(config) {
  const siteUrl = (config.site_url || 'https://example.com').replace(/\/+$/, '');
  const now = new Date().toISOString();
  const out = [];

  // Header
  out.push('# Bippsi Unified Agent Manifest\n');
  out.push('> **Spec:** Bippsi AI Standard v1.0 · <https://github.com/bigappstudiollc/bippsi/blob/main/ai-standard/SPEC.md>');
  out.push(`> **Generated:** ${now}\n`);
  out.push('> **For AI agents:** this single file replaces the need to fetch `llms.txt`, `AGENTS.md`, `agents.json`, `openapi.json`, and `/.well-known/mcp.json` separately. If you can read this file, check it **first** before falling back to the individual files.\n');
  out.push('---\n');

  // 1. Identity
  out.push('## 1. Identity\n');
  out.push('```json');
  out.push(jsonBlock(config.identity || {}));
  out.push('```\n');

  // 2. Agent access policy
  const p = config.agent_policy || {};
  out.push('## 2. Agent access policy\n');
  if (p.allowed?.length) {
    out.push('**Without authentication, agents MAY:**');
    p.allowed.forEach(i => out.push(`- ${i}`));
    out.push('');
  }
  if (p.denied?.length) {
    out.push('**Without authentication, agents MAY NOT:**');
    p.denied.forEach(i => out.push(`- ${i}`));
    out.push('');
  }
  if (p.auth_methods?.length) {
    out.push('**Authentication methods:**');
    p.auth_methods.forEach(i => out.push(`- ${i}`));
    out.push('');
  }
  if (p.rate_limits) {
    out.push('**Rate limits:**');
    out.push('```json');
    out.push(jsonBlock(p.rate_limits));
    out.push('```\n');
  }

  // 3. MCP server
  out.push('## 3. MCP (Model Context Protocol) server\n');
  if (!config.mcp) {
    out.push('This site does not currently run an MCP server.');
    out.push('Agents should fall through to the REST API (section 4) or legacy discovery (section 8).\n');
  } else {
    out.push('```json');
    out.push(jsonBlock(config.mcp));
    out.push('```\n');
    if (config.mcp.tools?.length) {
      out.push('### Available tools\n');
      for (const t of config.mcp.tools) {
        out.push(`#### \`${t.name || '?'}\`\n`);
        out.push(`${(t.description || '').trim()}\n`);
      }
    }
  }

  // 4. REST API
  out.push('## 4. REST API\n');
  if (!config.api) {
    out.push('This site does not publish a machine-readable REST API.\n');
  } else {
    const endpoints = typeof config.api === 'string' ? loadOpenApi(config.api) : config.api;
    out.push(`OpenAPI spec: ${siteUrl}/openapi.json\n`);
    out.push('```json');
    out.push(jsonBlock(endpoints));
    out.push('```\n');
  }

  // 5. Certification
  out.push('## 5. Certification status (this site)\n');
  if (!config.certification) {
    out.push('This site is not currently in any public certification directory.\n');
  } else {
    out.push('```json');
    out.push(jsonBlock(config.certification));
    out.push('```\n');
  }

  // 6. Licensing
  out.push('## 6. Content licensing for AI training\n');
  if (Array.isArray(config.licensing) && config.licensing.length) {
    config.licensing.forEach(l => out.push(`- ${l}`));
    out.push('');
  } else {
    out.push('No explicit licensing declared. Agents SHOULD NOT assume permission to use content for training or redistribution.\n');
  }

  // 7. Pages
  out.push('## 7. Page index\n');
  for (const pg of (config.pages || [])) {
    const url = pg.url || '/';
    const full = url.startsWith('http') ? url : `${siteUrl}${url}`;
    out.push(`- [${pg.title || url}](${full})`);
  }
  out.push('');

  // 8. Fallback files
  out.push('## 8. Fallback — individual discovery files\n');
  out.push("If an agent cannot parse this unified manifest, fall back to the following files at this site's root.");
  out.push('**If this manifest disagrees with any of them, this manifest is authoritative.**\n');
  out.push('| File | Purpose |');
  out.push('|---|---|');
  const fallback = config.fallback_files || defaultFallback();
  for (const f of fallback) out.push(`| \`${f.path || '?'}\` | ${f.purpose || ''} |`);
  out.push('');

  // 9. Spec metadata
  out.push('## 9. Spec metadata\n');
  out.push('```json');
  out.push(jsonBlock({
    spec: 'Bippsi AI Standard',
    spec_version: '1.0',
    spec_url: 'https://github.com/bigappstudiollc/bippsi/blob/main/ai-standard/SPEC.md',
    spec_source: 'https://bippsi.com/bippsi-standard',
    manifest_generated_at: now,
    manifest_version: '1.0',
    publisher: config.identity?.name || '',
  }));
  out.push('```\n');

  out.push('---\n');
  out.push('*This manifest follows the Bippsi AI Standard v1.0. Spec: <https://github.com/bigappstudiollc/bippsi>.*\n');

  return out.join('\n');
}

function jsonBlock(data) {
  return JSON.stringify(data, null, 2);
}

function loadOpenApi(path) {
  try {
    const spec = JSON.parse(readFileSync(path, 'utf8'));
    const base = (spec.servers?.[0]?.url || '').replace(/\/+$/, '');
    const out = [];
    for (const [p, ops] of Object.entries(spec.paths || {})) {
      for (const [method, op] of Object.entries(ops)) {
        if (!['get','post','put','patch','delete'].includes(method.toLowerCase())) continue;
        out.push({
          method:  method.toUpperCase(),
          url:     base + p,
          summary: op.summary || '',
          auth:    op.security?.length ? 'Bearer required' : 'none',
        });
      }
    }
    return out;
  } catch {
    return [];
  }
}

function defaultFallback() {
  return [
    { path: '/AGENTS.md',                  purpose: 'Narrative agent policy' },
    { path: '/llms.txt',                   purpose: 'LLM-friendly content index' },
    { path: '/agents.json',                purpose: 'Structured agent policy' },
    { path: '/openapi.json',               purpose: 'REST API spec (OpenAPI 3.1)' },
    { path: '/.well-known/mcp.json',       purpose: 'MCP server manifest' },
    { path: '/.well-known/ai-plugin.json', purpose: 'OpenAI plugin manifest' },
    { path: '/.well-known/security.txt',   purpose: 'Security disclosure (RFC 9116)' },
    { path: '/manifest.json',              purpose: 'PWA manifest' },
    { path: '/robots.txt',                 purpose: 'Crawler policy' },
    { path: '/sitemap.xml',                purpose: 'URL index' },
  ];
}

// ── CLI mode ──────────────────────────────────────────────────────────────────
// Run when this file is invoked directly: `node bippsi-unified.js config.json`
if (import.meta.url === `file://${process.argv[1].replace(/\\/g, '/')}`) {
  const configPath = argv[2];
  if (!configPath) {
    console.error('Usage: node bippsi-unified.js <config.json> > output.md');
    process.exit(1);
  }
  try {
    const config = JSON.parse(readFileSync(configPath, 'utf8'));
    process.stdout.write(renderBippsiUnified(config));
  } catch (err) {
    console.error(`Error: ${err.message}`);
    process.exit(1);
  }
}
